File: src/audio/sidetone.rs
//! Sidetone generator with phase accumulator and fade envelope
//!
//! Uses lookup table for sine wave, fixed-point phase accumulator
//! for frequency control, and linear fade for anti-click.

use super::lut::{SINE_LUT, LUT_SIZE};

/// Fade envelope state
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FadeState {
    /// Output is zero
    Silent,
    /// Ramping up (0 → full)
    FadeIn,
    /// Full amplitude output
    Sustain,
    /// Ramping down (full → 0)
    FadeOut,
}

/// Sidetone generator
///
/// Generates sine wave with configurable frequency and fade envelope.
/// Uses phase accumulator for frequency control (no floating point in hot path).
pub struct SidetoneGen {
    /// Phase accumulator (32-bit fixed point, top 8 bits = LUT index)
    phase: u32,
    /// Phase increment per sample (determines frequency)
    phase_inc: u32,
    /// Current fade state
    fade_state: FadeState,
    /// Current fade position (0 to fade_len)
    fade_pos: u16,
    /// Fade length in samples
    fade_len: u16,
}

impl SidetoneGen {
    /// Create new sidetone generator
    ///
    /// # Arguments
    /// * `freq_hz` - Sidetone frequency in Hz (typically 400-800)
    /// * `sample_rate` - Output sample rate in Hz (typically 8000)
    /// * `fade_samples` - Fade duration in samples (e.g., 40 for 5ms @ 8kHz)
    pub fn new(freq_hz: u32, sample_rate: u32, fade_samples: u16) -> Self {
        Self {
            phase: 0,
            phase_inc: Self::calc_phase_inc(freq_hz, sample_rate),
            fade_state: FadeState::Silent,
            fade_pos: 0,
            fade_len: fade_samples.max(1), // Avoid div by zero
        }
    }

    /// Calculate phase increment for target frequency
    ///
    /// phase_inc = (freq * 2^32) / sample_rate
    #[inline]
    fn calc_phase_inc(freq_hz: u32, sample_rate: u32) -> u32 {
        ((freq_hz as u64 * (1u64 << 32)) / sample_rate as u64) as u32
    }

    /// Update frequency (e.g., when config changes)
    #[inline]
    pub fn set_frequency(&mut self, freq_hz: u32, sample_rate: u32) {
        self.phase_inc = Self::calc_phase_inc(freq_hz, sample_rate);
    }

    /// Get current fade state
    #[inline]
    pub fn fade_state(&self) -> FadeState {
        self.fade_state
    }

    /// Generate next audio sample
    ///
    /// # Arguments
    /// * `key_down` - true if key is pressed (generate tone), false for silence
    ///
    /// # Returns
    /// Audio sample in i16 range
    #[inline]
    pub fn next_sample(&mut self, key_down: bool) -> i16 {
        // Update fade state and get amplitude multiplier
        let amplitude = self.update_fade(key_down);

        if amplitude == 0 {
            return 0;
        }

        // LUT lookup with phase accumulator
        let idx = (self.phase >> 24) as usize; // Top 8 bits = index (0-255)
        self.phase = self.phase.wrapping_add(self.phase_inc);

        // Get sample from LUT
        let sample = SINE_LUT[idx % LUT_SIZE];

        // Apply fade envelope
        ((sample as i32 * amplitude as i32) >> 16) as i16
    }

    /// Update fade state machine, returns amplitude (0-65535)
    #[inline]
    fn update_fade(&mut self, key_down: bool) -> u16 {
        match (self.fade_state, key_down) {
            // Silent + key down → start fade in
            (FadeState::Silent, true) => {
                self.fade_state = FadeState::FadeIn;
                self.fade_pos = 0;
                0
            }

            // Fade in + key down → continue fade
            (FadeState::FadeIn, true) => {
                self.fade_pos += 1;
                if self.fade_pos >= self.fade_len {
                    self.fade_state = FadeState::Sustain;
                    0xFFFF
                } else {
                    ((self.fade_pos as u32 * 0xFFFF) / self.fade_len as u32) as u16
                }
            }

            // Sustain + key down → full amplitude
            (FadeState::Sustain, true) => 0xFFFF,

            // Key up while fade in or sustain → start fade out
            (FadeState::FadeIn, false) | (FadeState::Sustain, false) => {
                self.fade_state = FadeState::FadeOut;
                self.fade_pos = self.fade_len;
                0xFFFF
            }

            // Fade out → continue fade
            (FadeState::FadeOut, _) => {
                if self.fade_pos == 0 {
                    self.fade_state = FadeState::Silent;
                    0
                } else {
                    self.fade_pos -= 1;
                    ((self.fade_pos as u32 * 0xFFFF) / self.fade_len as u32) as u16
                }
            }

            // Silent + key up → stay silent
            (FadeState::Silent, false) => 0,
        }
    }

    /// Reset generator state (e.g., after fault)
    pub fn reset(&mut self) {
        self.phase = 0;
        self.fade_state = FadeState::Silent;
        self.fade_pos = 0;
    }
}

File: components/keyer_audio/src/sidetone.c
/**
 * @file sidetone.c
 * @brief Sidetone generator implementation
 *
 * Phase accumulator with 256-entry sine LUT.
 * Digital fade envelope eliminates key clicks.
 */

#include "sidetone.h"
#include <assert.h>
#include <stddef.h>

/* Phase accumulator uses upper 8 bits for LUT index */
#define PHASE_SHIFT 24
#define PHASE_MASK  0xFF000000U

void sidetone_init(sidetone_gen_t *gen, uint32_t freq_hz, uint32_t sample_rate,
                   uint16_t fade_samples) {
    assert(gen != NULL);
    assert(sample_rate > 0);
    assert(fade_samples > 0);

    gen->phase = 0;
    gen->sample_rate = sample_rate;
    gen->fade_state = FADE_SILENT;
    gen->fade_pos = 0;
    gen->fade_len = fade_samples;

    /* Calculate phase increment: (freq * 2^32) / sample_rate */
    gen->phase_inc = (uint32_t)(((uint64_t)freq_hz << 32) / sample_rate);
}

void sidetone_set_frequency(sidetone_gen_t *gen, uint32_t freq_hz) {
    assert(gen != NULL);

    gen->phase_inc = (uint32_t)(((uint64_t)freq_hz << 32) / gen->sample_rate);
}

void sidetone_reset(sidetone_gen_t *gen) {
    assert(gen != NULL);

    gen->phase = 0;
    gen->fade_state = FADE_SILENT;
    gen->fade_pos = 0;
}

int16_t sidetone_next_sample(sidetone_gen_t *gen, bool key_down) {
    assert(gen != NULL);

    /* Update fade state machine */
    switch (gen->fade_state) {
        case FADE_SILENT:
            if (key_down) {
                gen->fade_state = FADE_IN;
                gen->fade_pos = 0;
            }
            break;

        case FADE_IN:
            if (!key_down) {
                gen->fade_state = FADE_OUT;
                /* Invert position for smooth transition */
                gen->fade_pos = gen->fade_len - gen->fade_pos;
            } else if (gen->fade_pos >= gen->fade_len) {
                gen->fade_state = FADE_SUSTAIN;
            } else {
                gen->fade_pos++;
            }
            break;

        case FADE_SUSTAIN:
            if (!key_down) {
                gen->fade_state = FADE_OUT;
                gen->fade_pos = 0;
            }
            break;

        case FADE_OUT:
            if (key_down) {
                gen->fade_state = FADE_IN;
                /* Invert position for smooth transition */
                gen->fade_pos = gen->fade_len - gen->fade_pos;
            } else if (gen->fade_pos >= gen->fade_len) {
                gen->fade_state = FADE_SILENT;
                gen->fade_pos = 0;
                return 0;
            } else {
                gen->fade_pos++;
            }
            break;
    }

    /* If silent, return 0 */
    if (gen->fade_state == FADE_SILENT) {
        return 0;
    }

    /* Get sine sample from LUT */
    uint8_t lut_idx = (uint8_t)(gen->phase >> PHASE_SHIFT);
    int32_t raw_sample = SINE_LUT[lut_idx];

    /* Advance phase */
    gen->phase += gen->phase_inc;

    /* Apply envelope */
    int32_t envelope;
    switch (gen->fade_state) {
        case FADE_IN:
            /* Linear ramp up */
            envelope = ((int32_t)gen->fade_pos * 32767) / gen->fade_len;
            break;

        case FADE_OUT:
            /* Linear ramp down */
            envelope = ((int32_t)(gen->fade_len - gen->fade_pos) * 32767) / gen->fade_len;
            break;

        case FADE_SUSTAIN:
        default:
            envelope = 32767;
            break;
    }

    /* Apply envelope and scale */
    int32_t sample = (raw_sample * envelope) >> 15;

    /* Clamp to int16_t range */
    if (sample > 32767) sample = 32767;
    if (sample < -32767) sample = -32767;

    return (int16_t)sample;
}
